cow_amm/cow_amm_lp_and_tvl_4047078.sql

```diff
@@ -0,0 +1,151 @@
+-- Computes the TVL, lp token total supply and lp token price of a CoW AMM pool over time
+-- Parameters
+--  {{token_a}} - either token of the desired uni pool
+--  {{token_b}} - other token of the desired uni pool
+--  {{start}} - date as of which the analysis should run
+
+-- Given that we might not have records every day in the source data (e.g. not every day the lp supply may change), 
+-- but still want to visualize development on a per day basis,  we create an auxiliary table with one record per 
+-- day between `start` and `now`
+with date_range as (
+    select t.day
+    from
+        unnest(sequence(
+            date(timestamp '{{start}}'),
+            date(now())
+        )) t (day) --noqa: AL01
+),
+
+-- Finds the CoW AMM pool address given tokens specified in query parameters (regardless of order) 
+cow_amm_pool as (
+    select
+        blockchain,
+        address,
+        token_1_address,
+        token_2_address
+    from query_3959044
+    where ((token_1_address = {{token_a}} and token_2_address = {{token_b}}) or (token_2_address = {{token_a}} and token_1_address = {{token_b}}))
+),
+
+-- per day lp token total supply changes of the CoW AMM pool by looking at burn/mint events
+lp_balance_delta as (
+    select
+        date(evt_block_time) as "day",
+        sum(case when "from" = 0x0000000000000000000000000000000000000000 then value else -value end) as lp_supply
+    from erc20_ethereum.evt_transfer
+    where
+        contract_address in (select address from cow_amm_pool)
+        and ("from" = 0x0000000000000000000000000000000000000000 or "to" = 0x0000000000000000000000000000000000000000)
+    group by 1
+),
+
+total_lp as (
+    select
+        day,
+        lp_supply,
+        sum(lp_supply) over (order by day) as total_lp
+    from lp_balance_delta
+),
+
+-- lp token total supply without date gaps
+lp_total_supply as (
+    select
+        day,
+        total_lp
+    from (
+        -- join full date range with potentially incomplete data. This results in many rows per day (all total supplies on or before that day)
+        -- rank() is then used to order join candidates by recency (rank = 1 is the latest lp supply)
+        select
+            date_range.day,
+            total_lp,
+            rank() over (partition by (date_range.day) order by lp.day desc) as latest
+        from date_range
+        inner join total_lp as lp
+            on date_range.day >= lp.day
+            -- performance optimisation: this assumes one week prior to start there was at least one lp supply change event
+            and lp.day >= (timestamp '{{start}}' - interval '7' day)
+    )
+    where latest = 1
+),
+
+-- Compute tvl by multiplying each day's closing price with net transfers in and out of the pool
+get_tvl as (
+    select
+        x.day,
+        sum(amount * price_close / pow(10, decimals)) as total_tvl
```

Comment:
This doesn't change anything compared to the logic before (you are still going into floating number before you are summing up all the in and out transfers thus creating imprecision).  

Instead, this needs to become (first summing the whole wei amounts and only then dividing by decimals):
```suggestion
        sum(amount) * price_close / pow(10, decimals) as total_tvl
```


If you want to see an example of the floating imprecision you can run the two python lines which give different output:

```python
(1/6)+(1/6)+(1/6)+(1/6)+(1/6)+(1/6)
(1+1+1+1+1+1)/6
```

Reply:
But this is not going to work in this query as price and decimals vary per token. So it either has to be done per day per token like this:

```
select day, sum(total_tvl)
from (
select 
        x.day,
        x.contract_address,
        decimals,
        price_close,
        sum(amount) * price_close / pow(10, decimals) as total_tvl
```
Or aggregated.